src/Native/ObjWriter/objwriter.cpp

```diff
@@ -317,6 +326,50 @@ ObjectWriter::GetSymbolRefExpr(const char *SymbolName,
 
 int ObjectWriter::EmitSymbolRef(const char *SymbolName,
                                 RelocType RelocationType, int Delta) {
+  if (GetTriple().getArch() == Triple::thumb)
+  {
+    const MCExpr *TargetExpr =
+      GetSymbolRefExpr(SymbolName, MCSymbolRefExpr::VK_None);
+    if (Delta != 0) {
+      TargetExpr = MCBinaryExpr::createAdd(TargetExpr,
+          MCConstantExpr::create(Delta, *OutContext), *OutContext);
+    }
+
+    MCDataFragment *DF = Streamer->getOrCreateDataFragment();
+    unsigned Offset = DF->getContents().size();
+    const MCExpr *OffsetExpr = MCConstantExpr::create(Offset, *OutContext);
+
+    switch (RelocationType) {
+    case RelocType::IMAGE_REL_BASED_REL32:
+      // If the fixup is pc-relative, we need to bias the value to be relative
+      // to the start of the field, not the end of the field
+      TargetExpr = MCBinaryExpr::createSub(
+          TargetExpr, MCConstantExpr::create(4, *OutContext), *OutContext);
+      Streamer->EmitValueImpl(TargetExpr, 4, SMLoc(), true);
+      return 4;
+
+    case RelocType::IMAGE_REL_BASED_HIGHLOW:
```

Comment:
Would it make sense to use the platform-neutral implementation for these instead of duplicating them?

Reply:
Initially, this patch was not made by me, it's by @mskvortsov, so I did not change anything. Well, I'll fix it.